### PR TITLE
Lookup plugin update: Update AWS SSM lookup to support getting parameters by path

### PR DESCRIPTION
**Pull request by mtb-xt**
##### SUMMARY
Update AWS SSM Parameter storage lookup plugin to support getting parameters by path

AWS SSM Parameter store lookup was added in PR https://github.com/ansible/ansible/pull/23460, but there were requests to add support to look up parameters by path. 

This is my attempt on that.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
plugins/lookup/ssm.py

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 3.6.2 (default, Jul 20 2017, 03:52:27) [GCC 7.1.1 20170630]
```

##### ADDITIONAL INFORMATION
Currently, the lookup plugin only gets a single parameter as in example below:
```
- debug: msg="{{ lookup('ssm', 'Hello', 'region=ap-southeast-2',  ) }}"
```

This PR adds these features:
You lookup the whole parameters hierarchy:
```
 - debug: msg="{{ lookup('ssm', '/Test', 'region=ap-southeast-2', 'bypath', 'recursive=true' ) }}
```
Would return:
```
ok: [localhost] => {
    "msg": [
        {
            "Name": "/Test/secure-test", 
            "Type": "SecureString", 
            "Value": "secure-test"
        }, 
        {
            "Name": "/Test/subpath/test", 
            "Type": "String", 
            "Value": "testsubpath"
        }, 
        {
            "Name": "/Test/test-string", 
            "Type": "String", 
            "Value": "test-string"
        }
    ]
}
```

You can loop through them:
```
         - name: Getting parameters in a loop
           debug: msg='key contains an item named {{item.Name }} with value {{item.Value}}'
           with_ssm:
             - '/Test'
             - 'region=ap-southeast-2'
             - 'bypath'
             - 'recursive=true'
```
would return:

```
TASK [Getting parameters in a loop] *********************************************************************
Tuesday 10 October 2017  16:22:44 +1300 (0:00:00.370)       0:00:01.035 ******* 
ok: [localhost] => (item={u'Type': u'SecureString', u'Name': u'/Test/secure-test', u'Value': u'secure-test'}) => {
    "item": {
        "Name": "/Test/secure-test", 
        "Type": "SecureString", 
        "Value": "secure-test"
    }, 
    "msg": "key contains an item named /Test/secure-test with value secure-test"
}
ok: [localhost] => (item={u'Type': u'String', u'Name': u'/Test/subpath/test', u'Value': u'testsubpath'}) => {
    "item": {
        "Name": "/Test/subpath/test", 
        "Type": "String", 
        "Value": "testsubpath"
    }, 
    "msg": "key contains an item named /Test/subpath/test with value testsubpath"
}
ok: [localhost] => (item={u'Type': u'String', u'Name': u'/Test/test-string', u'Value': u'test-string'}) => {
    "item": {
        "Name": "/Test/test-string", 
        "Type": "String", 
        "Value": "test-string"
    }, 
    "msg": "key contains an item named /Test/test-string with value test-string"
}
```

You can also ask the plugin to return short versions of parameter names (The part after the last slash, not the full path):
```
         - name: Getting parameters in a loop with names shorteting
           debug: msg='key contains an item named {{item.Name }} with value {{item.Value}}'
           with_ssm:
             - '/Test'
             - 'region=ap-southeast-2'
             - 'bypath'
             - 'shortnames'
             - 'recursive=true'
```
Would return:

```
TASK [Getting parameters in a loop with names shorteting] ***********************************************
Tuesday 10 October 2017  16:22:44 +1300 (0:00:00.371)       0:00:01.406 ******* 
ok: [localhost] => (item={u'Type': u'SecureString', u'Name': u'secure-test', u'Value': u'secure-test'}) => {
    "item": {
        "Name": "secure-test", 
        "Type": "SecureString", 
        "Value": "secure-test"
    }, 
    "msg": "key contains an item named secure-test with value secure-test"
}
ok: [localhost] => (item={u'Type': u'String', u'Name': u'test', u'Value': u'testsubpath'}) => {
    "item": {
        "Name": "test", 
        "Type": "String", 
        "Value": "testsubpath"
    }, 
    "msg": "key contains an item named test with value testsubpath"
}
ok: [localhost] => (item={u'Type': u'String', u'Name': u'test-string', u'Value': u'test-string'}) => {
    "item": {
        "Name": "test-string", 
        "Type": "String", 
        "Value": "test-string"
    }, 
    "msg": "key contains an item named test-string with value test-string"
}
```

Since you were involved in the other PR, would you be able to look at this @willthames ? 
Or does this need reviews from other maintainers?

### CODE REVIEW

File: lib/ansible/plugins/lookup/ssm.py [hunk 1]
```diff
@@ -73,6 +99,10 @@ def run(self, terms, variables, **kwargs):
                 if key == "region" or key == "aws_profile":
                     session[key] = value
 
+                # recurse or not
+                if key == "recursive" and value.lower() == "true":
```
Comment (willthames) [inline lib/ansible/plugins/lookup/ssm.py]: might want to use 

```
from ansible.module_utils.parsing.convert_bool import boolean
...

if key == "recursive" and boolean(value):
     ssm_dict['Recursive'] = True
...


for this, as people might expect to be able to use 'yes' or 'on'

**Comment by mtb-xt** [comment]
@ansibot  bot_status

File: lib/ansible/plugins/lookup/ssm.py [hunk 2]
```diff
@@ -86,13 +116,34 @@ def run(self, terms, variables, **kwargs):
...
+                response = client.get_parameters_by_path(**ssm_dict)
+                paramlist = list()
+                paramlist.extend(response['Parameters'])
+
+                # Manual pagination, since boto doesnt support it yet for get_parameters_by_path
+                while 'NextToken' in response:
+                    response = client.get_parameters_by_path(NextToken=response['NextToken'], **ssm_dict)
+                    paramlist.extend(response['Parameters'])
+
+                # shortify parameter names. yes, this will return duplicate names with different values.
```
Comment (willthames) [inline lib/ansible/plugins/lookup/ssm.py]: shorten, not shortify

**Review by willthames — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/lookup/ssm.py (hunk 1 above)
Comment (mtb-xt) [reply lib/ansible/plugins/lookup/ssm.py]: Thanks! I've changed both *recursive* and *decrypt*

File: lib/ansible/plugins/lookup/ssm.py (hunk 2 above)
Comment (mtb-xt) [reply lib/ansible/plugins/lookup/ssm.py]: Noted

**Comment by mtb-xt** [comment]
@willthames I've made the requested changes

**Comment by willthames** [comment]
This will need rebasing upon the rename now

**Review by willthames — APPROVED** [review]

**Comment by viper233** [comment]
I have a problem with lookups within the AWS environment as it seems impossible to pass STS tokens for cross account access. Seeing aws_ssm_parameter_store exists, wouldn't it be better to have aws_ssm_parameter_store_facts?

**Comment by mtb-xt** [comment]
@willthames I've screwed up rebasing this PR, so I opened another one - https://github.com/ansible/ansible/pull/31752 . Sorry about that :/

**Pull request closed without merging**